### PR TITLE
Store user and strategy stats

### PR DESCRIPTION
**Pull request by tanim0la**
Description
-----------

With this PR, user and strategy stats can now be stored in the Strategy Module contract.

Motivation and Context
----------------------

This stored information is important for the frontend, enabling users to track their assets.

How Has This Been Tested?
-------------------------

In `Join.t.sol`, I tested this by joining a strategy and then using the `getUserShareBalance` method in the Strategy Module to retrieve the protocol address.

Screenshots (if appropriate, particularly for frontend repos)
----------------------------

Types of changes
----------------

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

Checklist
---------

- [x] My pull request addresses exactly one patch/feature.
- [x] I have created a branch for this patch/feature.
- [x] Each individual commit in the pull request is meaningful. (Please rebase, removing useless commits if not)
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] My code follows the code style of this project.
- [ ] My change requires change(s) to the environment.
- [ ] Communication sent to affected teams

- [ ] Other information.

### CODE REVIEW

File: src/curators/engine.sol [hunk 1]
```diff
@@ -1 +1,85 @@
...
+                require(verifyResult(amountOut, _step.assetOut, _step.connector), "Invalid result");
+
+                // Update the strategy module
+            } catch Error(string memory reason) {
+                revert(string(abi.encodePacked("Step ", i, " failed: ", reason)));
+            }
+        }
+
+        // zero free for now
+        uint256 _fee = 0;
```
[the remark this answers is not in the document]
Comment (njokuScript) [reply src/curators/engine.sol:72]: Also, thinking we'll set a base curator fee of 10% of rewards/yields instead so we'll remove the fee param on create strategy.

**Review by njokuScript — CHANGES_REQUESTED** [review]

File: src/curators/strategy.sol [hunk 2]
```diff
@@ -117,6 +133,77 @@ contract Strategy {
         );
     }
 
+    function transferToken(address _token, uint256 _amount) public returns (bool) {
+        // check if amount equals or more than available balance first
+        return IERC20(_token).transfer(msg.sender, _amount);
+    }
```
Comment (njokuScript) [inline src/curators/strategy.sol:139]: what does this do? @tanim0la

File: src/protocols/common/constant.sol [hunk 3]
```diff
@@ -18,6 +18,19 @@ abstract contract Constants {
...
+    address internal constant ETH = 0x4200000000000000000000000000000000000006;
+    address internal constant USDC = 0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913;
+
+    // Oracle
+    address internal constant ORACLE = 0x333Cd307bd0d8fDB3c38b14eacC4072FF548176B;
+    address internal constant ENGINE = 0xe7D11A96aB3813D8232a0711D4fa4f60E2f50B19;
+
+    // Moonwell Basic contracts
+    address internal constant COMPTROLLER = 0xfBb21d0380beE3312B33c4353c8936a0F13EF26C;
+
```
Comment (njokuScript) [inline src/protocols/common/constant.sol:33]: can you add EUROc

File: src/curators/strategy.sol (hunk 2 above)
Comment (tanim0la) [reply src/curators/strategy.sol:139]: its purpose is to transfers tokens from the strategy module to the connector. still need to address necessary checks before the transfer is done

Comment (njokuScript) [reply src/curators/strategy.sol:139]: this will be used when exiting a strategy right?

Comment (njokuScript) [reply src/curators/strategy.sol:139]: can you also emit events after the join strategy is complete?

Comment (tanim0la) [reply src/curators/strategy.sol:139]: not only when exiting a strategy, the idea is the share amount gotten from each step is transferred to the strategy module and the balance is updated, then if the next step needs the share token from the previous step, the token is transferred to the connector.

check the engine contract, line 90

Comment (njokuScript) [reply src/curators/strategy.sol:139]: got it. this is line 90 - https://github.com/tanim0la/liquid-protocol/blob/feature/eng-19-strategy-module-development/src/curators/engine.sol#L90-L92

does it happen inside the update strategy func?

Comment (tanim0la) [reply src/curators/strategy.sol:139]: sorry i meant moonwell connector contract

Comment (tanim0la) [reply src/curators/strategy.sol:139]: here https://github.com/tanim0la/liquid-protocol/blob/69e9fb00a438ed8385b32a2c53c581bcb0c4c743/src/protocols/lending/base/moonwell/main.sol#L90

Comment (njokuScript) [reply src/curators/strategy.sol:139]: can you add the same  to both the morpho and aerodrome connectors? they are missing the transfer functions @tanim0la

Comment (tanim0la) [reply src/curators/strategy.sol:139]: sure, i'm not done with moonwell connectors, still need to implement more actions

Comment (njokuScript) [reply src/curators/strategy.sol:139]: got it

**Pull request merged**